src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt

```diff
@@ -84,3 +105,110 @@ fun getPolicyToTemplateMap(response: SearchResponse, xContentRegistry: NamedXCon
 @Suppress("UNCHECKED_CAST")
 fun <K, V> Map<K, V?>.filterNotNullValues(): Map<K, V> =
     filterValues { it != null } as Map<K, V>
+
+// get metadata from config index using doc id
+@Suppress("ReturnCount")
+suspend fun IndexMetadata.getManagedIndexMetaData(client: Client): ManagedIndexMetaData? {
+    try {
+        val getRequest = GetRequest(INDEX_MANAGEMENT_INDEX, indexUUID + "metadata")
+            .routing(this.indexUUID)
+        val getResponse: GetResponse = client.suspendUntil { get(getRequest, it) }
+        if (!getResponse.isExists || getResponse.isSourceEmpty) {
+            return null
+        }
+
+        return withContext(Dispatchers.IO) {
+            val xcp = XContentHelper.createParser(
+                NamedXContentRegistry.EMPTY,
+                LoggingDeprecationHandler.INSTANCE,
+                getResponse.sourceAsBytesRef, XContentType.JSON)
+            ManagedIndexMetaData.parseWithType(xcp, getResponse.id, getResponse.seqNo, getResponse.primaryTerm)
+        }
+    } catch (e: Exception) {
+        when (e) {
+            is IndexNotFoundException, is NoShardAvailableActionException -> {
+                log.error("Failed to get metadata because no index or shard not available")
+            }
+            else -> log.error("Failed to get metadata", e)
+        }
+
+        return null
+    }
+}
+
+/**
+ * multi-get metadata for indices
+ *
+ * @return list of metadata
+ */
+suspend fun Client.mgetManagedIndexMetadata(indices: List<Index>): List<ManagedIndexMetaData?> {
+    log.debug("trying to get back metadata for indices ${indices.map { it.name }}")
+
+    if (indices.isEmpty()) return emptyList()
+
+    val mgetRequest = MultiGetRequest()
+    indices.forEach {
+        mgetRequest.add(MultiGetRequest.Item(
+            INDEX_MANAGEMENT_INDEX, it.uuid + "metadata").routing(it.uuid))
+    }
+    var mgetMetadataList = mutableListOf<ManagedIndexMetaData?>()
+    try {
+        val response: MultiGetResponse = this.suspendUntil { multiGet(mgetRequest, it) }
+        mgetMetadataList = mgetResponseToList(response)
+    } catch (e: ActionRequestValidationException) {
+        log.info("No documents to get back metadata, ${e.message}")
+    }
+    return mgetMetadataList
+}
+
+/**
+ * transform multi-get response to list for ManagedIndexMetaData
+ */
+fun mgetResponseToList(mgetResponse: MultiGetResponse): MutableList<ManagedIndexMetaData?> {
+    val mgetList = mutableListOf<ManagedIndexMetaData?>()
+    mgetResponse.responses.forEach {
+        if (it.response != null && !it.response.isSourceEmpty) {
+            val xcp = contentParser(it.response.sourceAsBytesRef)
+            mgetList.add(ManagedIndexMetaData.parseWithType(
+                    xcp, it.response.id, it.response.seqNo, it.response.primaryTerm))
+        } else {
+            mgetList.add(null)
```

Comment:
Why adding null? Is it used?

Reply:
it's a list... the order is the clusterstate.metadata.indices

Reply:
I don't think I understand. Why are you creating a list of `[ManagedIndexMetadata, ManagedIndexMetadata, null, ManagedIndexMetadata, null, null]` etc.?

Are you preserving the indexes of the items on purpose for something? From what I see you're using here:

https://github.com/opendistro-for-elasticsearch/index-management/pull/280/files#diff-81b1f0c27a04a98265f20617842e0207f01e24f0774e0a0de7db49896f63a554R221

Which is just filtering out the nulls anyways. So my Q is why add these here?

Reply:
also used in change and retry API